AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy

```diff
@@ -0,0 +1,662 @@
+    var timestamp? := Time.GetCurrentTimeStamp();
+    var timestamp :- timestamp?
+    .MapFailure(e => Types.KeyStoreAdminException(message := e));
+
+    var mutationLockUUID? := UUID.GenerateUUID();
+    var mutationLockUUID :- mutationLockUUID?
+    .MapFailure(e => Types.KeyStoreAdminException(message := "Could not generate UUID for Mutation Lock. " + e));
+
+    var customEncryptionContext
+      := map k <- activeItem.EncryptionContext
+                  // This pull everything that is not in our restricted list.
+             | k !in Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES
+      :: k := activeItem.EncryptionContext[k];
+
+    var prefixedTerminalCustomEC?: Option<KeyStoreTypes.EncryptionContextString> := None;
+    var defixedTerminalCustomEC?: Option<KeyStoreTypes.EncryptionContextString> := None;
+    if (input.mutations.terminalEncryptionContext.Some?) {
+      var prefixedTerminalCustomEC := map k <- input.mutations.terminalEncryptionContext.value
+        :: Structure.ENCRYPTION_CONTEXT_PREFIX + k := input.mutations.terminalEncryptionContext.value[k];
+      prefixedTerminalCustomEC? := Some(prefixedTerminalCustomEC);
+      defixedTerminalCustomEC? := Some(input.mutations.terminalEncryptionContext.value);
+    }
+
+    var MutationToApply := StateStrucs.MutationToApply(
+      Identifier := input.branchKeyIdentifier,
+      Original := StateStrucs.MutableProperties(
+        kmsArn := KeyStoreTypes.kmsKeyArn(activeItem.KmsArn),
+        customEncryptionContext := customEncryptionContext
+      ),
+      Terminal := StateStrucs.MutableProperties(
+        kmsArn := KeyStoreTypes.kmsKeyArn(input.mutations.terminalKmsArn.UnwrapOr(activeItem.KmsArn)),
+        customEncryptionContext := prefixedTerminalCustomEC?.UnwrapOr(customEncryptionContext)
+        // customEncryptionContext := input.mutations.terminalEncryptionContext.UnwrapOr(customEncryptionContext)
+      ),
+      ExclusiveStartKey := None,
+      UUID := Some(mutationLockUUID),
+      CreateTime := timestamp
+    );
+
+      // -= BEGIN Version Active Branch Key
+      // --= Validate Active Branch Key
+    :- VerifyEncryptedHierarchicalKey(
+      item := activeItem,
+      keyManagerStrategy := keyManagerStrategy
+    );
+
+      // -= Assert Beacon Key is in Original
+    :- Need(
+      readItems.beaconItem.EncryptionContext
+      ==
+      Structure.ReplaceMutableContext(
+        readItems.beaconItem.EncryptionContext,
+        readItems.beaconItem.KmsArn,
+        MutationToApply.Original.customEncryptionContext),
+      Types.UnexpectedStateException(
+        message :=
+          "Beacon Item is not in the Original State!"
+          + " For Initialize Mutation to succeed, the ACTIVE & Beacon Key MUST be in the original state."
+      ));
+    // -= Validate Beacon Key
+    :- VerifyEncryptedHierarchicalKey(
+      item := readItems.beaconItem,
+      keyManagerStrategy := keyManagerStrategy
+    );
```

Comment:
Issue/Blocking: This is not needed.
The KMS ReEncrypt call verifies the integrity of the Beacon Key.
```suggestion
```